docs: add scale and HA sizing guide. index.md: extend. (#466)
kubara-io/kubara: docs/content/2_concepts/scale_and_ha.md

Reviewer: You need to add the new page to the mkdocs.yaml navigation block

# Scale and High Availability

## How far can kubara scale, and how many clusters can kubara handle?

This is one of the questions we have received several times from the community. To answer it, we ran scale tests with up to 1,000 vclusters and used the results to define practical sizing guidelines for kubara users.

The key question is not simply how many clusters or applications a single hub can manage. What matters more in practice is how many cached objects one hub can handle before resource limits need to be increased or an HA setup becomes necessary.

To make this easier to understand, we created the simplified view below. kubara currently relies on Argo CD as its GitOps engine, which means that kubara fleet management scales as far as Argo CD can scale. However, Argo CD is not the only component that needs to be considered. Platform components bootstrapped through kubara also need to be sized accordingly.

## Overview

> *Note: These tests were performed with kubara versions `0.7.0` and `0.8.0`, Argo CD versions `v3.3.7` and `v3.3.9`, and STACKIT Kubernetes Engine (SKE) version `1.35.4` between May and June 2026.*

Reviewer: ```suggestion
!!! note

    These tests were performed with kubara versions `0.7.0` and `0.8.0`, Argo CD versions `v3.3.7` and `v3.3.9`, and STACKIT Kubernetes Engine (SKE) version `1.35.4` between May and June 2026.
```

Reviewer: proper rendering of mkdocs-material notes

> ![image](../images/kubara-t-shirt-sizes.png)

Reviewer: move below the next section

Author: 🙌

Author: @tuunit Done. Please take a quick look at it.

> As the test results show, the limiting factor is not the number of vclusters or applications by itself. vclusters alone do not add much weight to Argo CD, as they are represented as Secrets and do not create significant load on their own. Applications, however, can vary significantly in size. A small nginx deployment creates very little load, while something like `kube-prometheus-stack` creates many Kubernetes objects. For that reason, judging scalability only by the number of vclusters or applications is misleading.

Reviewer: ```suggestion
As the test results show, the limiting factor is not the number of vclusters or applications by itself. vclusters or k8s clusters in general do not add much weight to Argo CD, as they are represented as Secrets and do not create significant load on their own. Applications, however, can vary significantly in size. A small nginx deployment creates very little load, while something like `kube-prometheus-stack` creates many Kubernetes objects. For that reason, judging scalability only by the number of clusters or applications is misleading.
```

The relationship between vclusters and applications is important, and it is not linear. Running 50 applications on one vcluster is not the same as running 50 applications across 50 vclusters, because more metadata and more live-state data need to be cached. For our tests, we used six applications per vcluster to create a balanced and realistic weighting.

> Based on this assumption, we found that the number of cached objects in the Argo CD Application Controller matters more than the number of clusters or applications alone. This document does not refer to Kubernetes objects stored in etcd, but to objects cached by the Argo CD Application Controller. The following three t-shirt sizes should give you a sense of kubara’s scaling boundaries and when resource limits or HA mode should be considered.

Reviewer: ```suggestion
The relationship between clusters and applications is important, and it is not linear. Running 50 applications on one cluster is not the same as running 50 applications across 50 cluster, because more metadata and more live-state data needs to be cached. For our tests, we used six applications per cluster to create a balanced and realistic weighting.

Based on this assumption, we found that the number of cached objects in the Argo CD Application Controller matters more than the number of clusters or applications alone. This document does not refer to Kubernetes objects stored in etcd, but to objects cached by the Argo CD Application Controller. The following three t-shirt sizes should give you a sense of kubara’s scaling boundaries and when resource limits or HA mode should be considered.
```

Reviewer: You have written the following three t-shirt sizes in the last paragraph of this section. Therefore shouldn't the image be below this paragraph instead of above?

> ### Argo CD Sizing – Component Configuration per Tier

Reviewer: ```suggestion
### Argo CD Sizing – Component Configuration per T-Shirt size
```

|                               | CPU (request / limit)    | 100m / -      | 250m / -     | 500m / -     |
|                               | Memory (request / limit) | 256Mi / 256Mi | 512Mi / 2Gi  | 4Gi / 8Gi    |
| **ApplicationSet Controller** | Replicas                 | 1             | 1            | 2            |
|                               | CPU (request / limit)    | 100m / -      | 250m / -     | 500m / -     |
|                               | Memory (request / limit) | 512Mi / 512Mi | 512Mi / 1Gi  | 2Gi / 4Gi    |

When moving from S to M or L, Argo CD should not be scaled in isolation. From M onward, platform components such as Kyverno, cert-manager, and others should be reviewed and scaled together with Argo CD.

This sizing is the no-fine-tuning baseline. It should make clear which resources start to matter when kubara, Argo CD, and the platform components are scaled together.

After this baseline is understood, Argo CD can be tuned further. The options below are the most important levers we used to improve scaling behavior and reduce avoidable pressure on the controller, repo-server, Redis, and Kubernetes API.

> *Note: HA is not just about replicas; it also requires anti-affinity, PodDisruptionBudgets, topology spread constraints, and failure-domain aware placement.*

Reviewer: ```suggestion
!!! note

    HA is not just about replicas; it also requires anti-affinity, PodDisruptionBudgets, topology spread constraints, and failure-domain aware placement.*
```

**Reduce Redis pressure with [`ARGOCD_APPLICATION_TREE_SHARD_SIZE`](https://argo-cd.readthedocs.io/en/stable/operator-manual/high_availability/#argocd-application-controller):** split large application trees across multiple Redis keys. A useful starting value is `100`; the default is `0`, which disables application-tree sharding. Also verify that `redis.compression` is set to `gzip`, which is the default, rather than treating compression as a separate feature that needs to be enabled.

**Raise Kubernetes client limits deliberately:** tune `ARGOCD_K8S_CLIENT_QPS` and `ARGOCD_K8S_CLIENT_BURST` as described in the Argo CD [application-controller scaling guidance](https://argo-cd.readthedocs.io/en/stable/operator-manual/high_availability/#argocd-application-controller). Otherwise, controller shards managing many clusters can be throttled by their own Kubernetes client.

> **Smooth cluster-cache rebuilds:** tune `ARGOCD_CLUSTER_CACHE_LIST_PAGE_SIZE` and `ARGOCD_CLUSTER_CACHE_LIST_SEMAPHORE` using the Argo CD [cluster-cache recommendations](https://argo-cd.readthedocs.io/en/stable/operator-manual/high_availability/#argocd-application-controller) to reduce memory spikes during startup and cache resyncs across many vClusters.

Reviewer: ```suggestion
**Smooth cluster-cache rebuilds:** tune `ARGOCD_CLUSTER_CACHE_LIST_PAGE_SIZE` and `ARGOCD_CLUSTER_CACHE_LIST_SEMAPHORE` using the Argo CD [cluster-cache recommendations](https://argo-cd.readthedocs.io/en/stable/operator-manual/high_availability/#argocd-application-controller) to reduce memory spikes during startup and cache resyncs across many clusters.
```
